modules/javafx.graphics/src/main/java/com/sun/javafx/application/preferences/PlatformPreferences.java

```diff
@@ -0,0 +1,237 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+/**
+ * Contains the implementation of a read-only map of platform preferences.
+ * <p>
+ * When the operating system signals that a preference has changed, the mappings are updated
+ * by calling the {@link #update(Map)} method.
+ */
+public class PlatformPreferences extends AbstractMap<String, Object> implements Platform.Preferences {
+
+    /**
+     * Contains mappings from platform-specific keys to well-known keys, which are used
+     * in the implementation of the property-based API in {@link PreferenceProperties}.
+     */
+    final Map<String, String> platformKeyMappings;
+    final Map<String, Object> effectivePreferences = new HashMap<>();
+    final Map<String, Object> unmodifiableEffectivePreferences = Collections.unmodifiableMap(effectivePreferences);
+    final PreferenceProperties properties = new PreferenceProperties(this);
+
+    private final List<InvalidationListener> invalidationListeners = new CopyOnWriteArrayList<>();
+    private final List<MapChangeListener<? super String, Object>> mapChangeListeners = new CopyOnWriteArrayList<>();
+
+    public PlatformPreferences(Map<String, String> platformKeyMappings) {
+        this.platformKeyMappings = platformKeyMappings;
+    }
+
+    @Override
+    public Set<Entry<String, Object>> entrySet() {
+        return unmodifiableEffectivePreferences.entrySet();
+    }
+
+    @Override
+    public Object remove(Object key) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void addListener(InvalidationListener listener) {
+        invalidationListeners.add(listener);
+    }
+
+    @Override
+    public void removeListener(InvalidationListener listener) {
+        invalidationListeners.remove(listener);
+    }
+
+    @Override
+    public void addListener(MapChangeListener<? super String, ? super Object> listener) {
+        mapChangeListeners.add(listener);
+    }
+
+    @Override
+    public void removeListener(MapChangeListener<? super String, ? super Object> listener) {
+        mapChangeListeners.remove(listener);
+    }
+
+    @Override
+    public <T> Optional<T> getValue(String key, Class<T> type) {
+        Objects.requireNonNull(key, "key cannot be null");
+        Objects.requireNonNull(type, "type cannot be null");
+        Object value = effectivePreferences.get(key);
+
+        if (value == null) {
+            return Optional.empty();
+        }
+
+        if (type.isInstance(value)) {
+            @SuppressWarnings("unchecked")
+            T v = (T)value;
+            return Optional.of(v);
+        }
+
+        throw new IllegalArgumentException(
+            "Incompatible types: requested = " + type.getName() +
+            ", actual = " + value.getClass().getName());
+    }
+
+    @Override
+    public Optional<Integer> getInteger(String key) {
```

Comment:
Q: is it possible to encounter an unsigned 64 bit value? 
Should we allow for BigInteger?

Reply:
IMHO this can be added when there's a use case.

Also, an int can support all values of an unsigned variant, you just need to be careful when interpreting it (ie. do `long unsigned = i & 0xffffffffL` first).

Reply:
I also think that that's an addition for another day.